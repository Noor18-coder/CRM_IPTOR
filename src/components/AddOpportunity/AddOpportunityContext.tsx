import * as React from 'react';
import { AddOpportunityDefaultParams, CustomerDetailsContactsGroupItem} from '../../helpers/Api/models';

export interface AddOpportunityContextInterface {
  customerId?: string,
  customerName?:string;
}

export const Context = React.createContext<AddOpportunityContextInterface | null>(null);

export const AddOpportunityContextProvider = Context.Provider;
  
export const AddOpportunityContextConsumer = Context.Consumer;