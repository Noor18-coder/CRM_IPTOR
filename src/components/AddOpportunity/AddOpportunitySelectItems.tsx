import React, { useCallback } from 'react';
import { Dispatch } from 'redux';
import { Image } from 'react-bootstrap';
import { useDispatch, useSelector } from 'react-redux';
import { AppState } from '../../store/store';
import { setOpportunityLoader, setOpportunityProducts } from '../../store/AddOpportunity/Actions';

// import ImageConfig from '../../config/ImageConfig';
// import ItemsList from './ItemsList';
import Item from '../../helpers/Api/Items';
import * as models from '../../helpers/Api/models';
import { Constants } from '../../config/Constants';

import VectorImg from '../../assets/images/check_circle.svg';

interface Props {
  createOpportunity: () => void;
  changeStep: (num: number) => void;
}

const AddOpportunitySelectItems: React.FC<Props> = ({ createOpportunity, changeStep }) => {
  const state: AppState = useSelector((appState: AppState) => appState);
  const dispatch: Dispatch<any> = useDispatch();
  const [searchText, setSearchText] = React.useState<string>('');
  const [items, setItems] = React.useState<models.Item[]>([]);
  const [hasMoreRows, setHasMoreRows] = React.useState<boolean>(false);
  const [pageNumber, setPageNumber] = React.useState<number>(0);

  // To handle pagination.
  const observer = React.useRef<IntersectionObserver>();
  const lastOpptyElement = useCallback(
    (node: any) => {
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver((entries) => {
        if (entries[0].isIntersecting) {
          setPageNumber(pageNumber + 1);
        }
      });
      if (node) observer.current.observe(node);
    },
    [hasMoreRows]
  );

  React.useEffect(() => {
    setHasMoreRows(false);
    fetchItems(pageNumber, searchText);
  }, [pageNumber]);

  const onNextButtonClick = () => {
    createOpportunity();
  };

  const fetchItems = async (currentPage: number, searchstr: string) => {
    dispatch(setOpportunityLoader(true));
    try {
      const offSet = currentPage * Constants.ADD_OPPTY_PRODUCTS_LOAD_LIMIT;
      const response = await Item.get(searchstr, offSet, Constants.ADD_OPPTY_PRODUCTS_LOAD_LIMIT);
      let totalCount;

      if (response && response.data && response.data.items) {
        if (currentPage) {
          setItems((prevItems) => [...prevItems, ...response.data.items]);
        } else {
          setItems(response.data.items);
        }

        if (response.control && response.control.total) {
          totalCount = response.control.total;
        } else {
          totalCount = response.data?.items.length;
        }

        if (offSet + response.data?.items.length < totalCount) {
          setHasMoreRows(true);
        }
      }
    } catch (e: any) {
      console.error(e);
    } finally {
      dispatch(setOpportunityLoader(false));
    }
  };

  const searchStart = (event: React.ChangeEvent<HTMLInputElement>) => {
    const str = event.target.value;
    setSearchText(str);
    if (str.length === 0) {
      fetchItems(0, '');
    }
  };

  const searchOpportunity = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === 'Enter') {
      setPageNumber(0);
      fetchItems(0, searchText);
    }
  };

  const onSelect = (str: string) => {
    const selectedItems: models.Item[] = [...state.addOpportunity.products];
    const find: models.Item | undefined = selectedItems.find((item: models.Item) => {
      return item.item === str;
    });

    if (find) {
      const newArray = selectedItems.filter((itemObj: models.Item) => {
        return itemObj.item !== str;
      });
      dispatch(setOpportunityProducts(newArray));
    } else {
      const newItem: models.Item | undefined = items.find((itemObj: models.Item) => {
        return itemObj.item === str;
      });
      if (newItem) dispatch(setOpportunityProducts([...selectedItems, newItem]));
    }
  };

  const isSelected = (item: string) => {
    let check = false;
    const selectedItems: models.Item[] = [...state.addOpportunity.products];
    const find = selectedItems?.find((selectedItem: models.Item) => {
      return selectedItem.item === item;
    });
    if (find) {
      check = true;
    }
    return check;
  };

  return (
    <>
      <div className="opportunity-step-circles">
        <ul className="list-inline step-circles">
          <li className="list-inline-item circle-stepone steps" role="presentation" onClick={() => changeStep(1)}>
            <span className="num">1</span>
          </li>
          <li className="list-inline-item circle-steptwo steps" role="presentation" onClick={() => changeStep(2)}>
            <span className="num">2</span>
          </li>
          <li className="list-inline-item circle-stepthree steps active" role="presentation">
            <span className="num">3</span>
          </li>
        </ul>
      </div>

      <div className="opportunity-forms thirdstep form-top">
        <div className="selected-products">
          <div className="thirdstep-tag-hscroll">
            {state.addOpportunity.products.map((item: models.Item) => {
              return (
                <div className="item-description" onClick={() => onSelect(item.item)} onKeyDown={() => onSelect(item.item)} role="presentation">
                  {item.description}
                </div>
              );
            })}
          </div>
        </div>
        <div className="">
          <div className="steps-three-forms">
            <div className="form-group oppty-form-elements">
              <input
                type="text"
                className="form-control search-ipbox"
                placeholder="Search Item"
                onChange={searchStart}
                onKeyPress={searchOpportunity}
              />
            </div>

            <div className="radiobtn-collection oppty-form-elements">
              <p className="title">Select Product and Modules</p>

              <div className="product-items-container">
                <ul className="opptytype-list-item">
                  {items.map((obj: models.Item, index: number) => {
                    if (index + 1 === items.length) {
                      return (
                        <li role="presentation" key={obj.item} ref={lastOpptyElement} onClick={() => onSelect(obj.item)}>
                          <div className="company-container">
                            <div className="center">
                              <div className="test">{obj.description}</div>
                              <Image
                                className="company-selection-img"
                                style={{ display: isSelected(obj.item) ? 'block' : 'none' }}
                                src={VectorImg}
                                alt="company"
                                title="company"
                              />
                            </div>
                          </div>
                        </li>
                      );
                    } else {
                      return (
                        <li role="presentation" key={obj.item} onClick={() => onSelect(obj.item)}>
                          <div className="company-container">
                            <div className="center">
                              <div className="test">{obj.description}</div>
                              <Image
                                className="company-selection-img"
                                style={{ display: isSelected(obj.item) ? 'block' : 'none' }}
                                src={VectorImg}
                                alt="company"
                                title="company"
                              />
                            </div>
                          </div>
                        </li>
                      );
                    }
                  })}
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div className="step-nextbtn-with-arrow stepsone-nxtbtn">
        <button type="button" className="stepone-next-btn done" onClick={onNextButtonClick}>
          Done
        </button>
      </div>
    </>
  );
};

export default AddOpportunitySelectItems;
