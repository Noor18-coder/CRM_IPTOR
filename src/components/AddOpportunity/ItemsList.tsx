import React from "react";
import { Item } from "../../helpers/Api/models";
import VectorImg from "../../assets/images/check_circle.svg";
import { Image} from "react-bootstrap";

export interface Props {
  items : Item[];
  doClick: (key: string) => void;
  selected?:string[]
}

const ItemsList: React.FC<Props> = ({ items, doClick, selected }) => {

  const isSelected = (item:string) => {
    let check = false;
    const find =  selected?.find((selectedItem:string) => { return selectedItem === item });
    if(find?.length){
      check = true;
    }
    return check;
  }

  return (
    <ul className="opptytype-list-item">
      {items.map((obj:Item) => {
        return (
          <li key={obj.item}  onClick={() => doClick(obj.item)}>
            <div className={"company-container"}>
              <div className={"center"}>
                <div className={"test"}>{obj.description}</div>
                <Image
                  className={"company-selection-img"}
                  style={{ display: isSelected(obj.item) ? "block" : "none" }}
                  src={VectorImg}
                  alt="company"
                  title="company"
                ></Image>
              </div>
            </div>
          </li>
        );
      })}
    </ul>
  );
};

export default ItemsList;
