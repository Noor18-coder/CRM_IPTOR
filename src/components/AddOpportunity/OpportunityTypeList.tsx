import React from "react";
import { OpportunityType } from "../../helpers/Api/models";
import VectorImg from "../../assets/images/check_circle.svg";
import { Image} from "react-bootstrap";

export interface Props {
  opptyTypes : OpportunityType[];
  doClick: (key: string) => void;
  selected?:string
}

export const OpportunityTypeList: React.FC<Props> = ({ opptyTypes, doClick, selected }) => {
  return (
    <ul className="opptytype-list-item">
      {opptyTypes.map((obj:OpportunityType) => {
        return (
          <li key={obj.description}  onClick={() => doClick(obj.description)}>
            <div
              className={"company-container"}
              key={obj.description}
            >
              <div className={"center"}>
                <div className={"test"}>{obj.oppRecordType}</div>
                <Image
                  className={"company-selection-img"}
                  style={{ display: obj.description  === selected ? "block" : "none" }}
                  src={VectorImg}
                  alt="company"
                  title="company"
                ></Image>
              </div>
            </div>
          </li>
        );
      })}
    </ul>
  );
};
