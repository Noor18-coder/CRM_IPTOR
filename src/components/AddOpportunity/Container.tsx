import React from 'react';
import { useMediaQuery } from 'react-responsive';
import { useSelector } from 'react-redux';

import Drawer from '@material-ui/core/Drawer';
import AddOpportunity from './AddOpportunity';
import Loader from '../Shared/Loader/Loader';
import { AppState } from '../../store/store';
import { AddOpportunityContextProvider, AddOpportunityContextInterface } from './AddOpportunityContext';

const Container: React.FC<AddOpportunityContextInterface> = ({ customerId, customerName }) => {
  const state: AppState = useSelector((appState: AppState) => appState);

  const isMobile = useMediaQuery({ maxWidth: 767 });
  const isTablet = useMediaQuery({ minWidth: 768, maxWidth: 991 });

  const customerData: AddOpportunityContextInterface = {
    customerId,
    customerName,
  };

  return (
    <>
      <AddOpportunityContextProvider value={customerData}>
        {state.addOpportunity.loader ? <Loader /> : null}
        {state.addOpportunity.loader ? <Loader /> : null}
        {isMobile || isTablet ? (
          state.addOpportunity.addOpptyWindowActive ? (
            <AddOpportunity />
          ) : null
        ) : (
          <Drawer anchor="right" open={state.addOpportunity.addOpptyWindowActive}>
            <AddOpportunity />
          </Drawer>
        )}
      </AddOpportunityContextProvider>
    </>
  );
};

export default Container;
