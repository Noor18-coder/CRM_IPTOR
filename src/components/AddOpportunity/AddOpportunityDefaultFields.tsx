import React from 'react';
import { Dispatch } from 'redux';
import { useSelector, useDispatch } from 'react-redux';
import AsynSearchInput from '../Shared/Search/AsyncSearchInput';

import { AppState } from '../../store/store';
import ImageConfig from '../../config/ImageConfig';
import * as models from '../../helpers/Api/models';
import { OpportunityTypeList } from './OpportunityTypeList';
import { Context } from './AddOpportunityContext';
import i18n from '../../i18n';

import { saveOpportunityParams, setOpportunityContacts, setAddOpportunityError } from '../../store/AddOpportunity/Actions';
import { getOpportunityTypes, saveOpportunityStages } from '../../store/InitialConfiguration/Actions';
import CustomerList from '../../helpers/Api/CustomerList';
import CustomerDetailsApi from '../../helpers/Api/CustomerDetailsApi';

const regex = /^-?\d+\.?\d*$/;

interface Props {
  changeStep: (num: number) => void;
}

interface ErrorMessages {
  desc?: string;
  customer?: string;
  stage?: string;
  oppRecordType?: string;
}

const AddOpportunityDefaultFields: React.FC<Props> = ({ changeStep }) => {
  const state: AppState = useSelector((appState: AppState) => appState);
  // const opportunity: models.AddOpportunityDefaultParams = state.addOpportunity.opportunityDefaultParams;
  const dispatch: Dispatch<any> = useDispatch();
  const [selectedOpportunityType, selectOpportunityType] = React.useState('');
  const selectedContacts = state.addOpportunity.contacts;
  const [customerContacts, setCustomerContacts] = React.useState<models.CustomerDetailsContactsGroupItem[]>([]);
  const [opportunity, setOpportunityField] = React.useState<models.AddOpportunityDefaultParams>();
  const contextValue = React.useContext(Context);
  const [nextButtonEnabled, setNextButtonEnabled] = React.useState<boolean>(true);
  const [errors, setErrorMessages] = React.useState<ErrorMessages>();

  const searchCustomers = async (key: string): Promise<models.BusinessPartnerListItem[]> => {
    const data = await CustomerList.get(key, 20, 0);
    return data.data.items;
  };

  const loadCustomerContacts = async (customerId: string) => {
    const customerContactsData = await CustomerDetailsApi.getAllContactDetails(customerId);
    setCustomerContacts(customerContactsData);
  };

  const onSearchItemSelect = async (data: any) => {
    if (data && data.length) {
      const selectItem: models.BusinessPartnerListItem = data[0];
      setOpportunityField({
        ...opportunity,
        customer: selectItem.businessPartner,
        customerName: selectItem.description,
      });
      setErrorMessages({
        ...errors,
        customer: '',
      });
      loadCustomerContacts(selectItem.businessPartner);
    } else {
      setErrorMessages({
        ...errors,
        customer: i18n.t('blankFieldError'),
      });
    }
  };

  const onInputValueChange = (e: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
    setOpportunityField({
      ...opportunity,
      [e.currentTarget.id]: e.currentTarget.value,
    });
  };

  const onOpportunityTypeSelect = (type: string) => {
    setOpportunityField({
      ...opportunity,
      oppRecordType: type,
    });
    selectOpportunityType(type);
  };

  const isNumeric = (value: any) => {
    if (value && !value.match(regex)) {
      return false;
    }
    return true;
  };

  const validateStepSecond = () => {
    const type = opportunity?.oppRecordType;
    const { currency, endDate, estimatedValue, forecastCategory, area } = state.addOpportunity.opportunityDefaultParams;
    if (
      type &&
      currency &&
      endDate &&
      estimatedValue &&
      forecastCategory &&
      area &&
      isNumeric(opportunity?.estimatedValue) &&
      checkIfAllFieldsHasValue(type)
    ) {
      return true;
    } else {
      return false;
    }
  };

  const checkIfAllFieldsHasValue = (type: string) => {
    let check = true;
    const { attributes } = state.addOpportunity;
    const selectedOpportunityRecordType = state.enviornmentConfigs.crmOpportunityTypes.find((obj: models.OpportunityType) => {
      return obj.oppRecordType === type;
    });
    const fields = selectedOpportunityRecordType?.MANDATORY_FIELDS || [];

    fields.forEach((field: string) => {
      const currField = attributes.find((obj: models.UserDefinedFieldReduxParams) => {
        return obj.attributeType === field;
      });
      if (currField && (currField.attributeValue || currField.attributeValueD)) {
        if (currField.valueFormat === 'N' && !isNumeric(currField.attributeValue)) {
          check = false;
        }
      } else {
        check = false;
      }
    });
    return check;
  };

  const onNextButtonClick = () => {
    changeStep(2);
    dispatch(saveOpportunityParams(opportunity));
  };

  const onClickStep2 = () => {
    if (!nextButtonEnabled) {
      dispatch(saveOpportunityParams(opportunity));
      changeStep(2);
    } else {
      dispatch(setAddOpportunityError(i18n.t('mandatoryFieldsMsg')));
    }
  };

  const onClickStep3 = () => {
    const valid = validateStepSecond();
    if (!nextButtonEnabled && valid) {
      dispatch(saveOpportunityParams(opportunity));
      changeStep(3);
    } else {
      const error = nextButtonEnabled ? i18n.t('mandatoryFieldsMsg') : i18n.t('mandatoryFieldStep2');
      dispatch(setAddOpportunityError(error));
    }
  };

  const onBlur = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { id, value } = e.currentTarget;
    const inputElement = document.getElementById(id) as HTMLInputElement;
    if (value.length === 0) {
      setErrorMessages({
        ...errors,
        [id]: i18n.t('blankFieldError'),
      });
      inputElement.style.border = '1px solid #ED2024';
    } else {
      setErrorMessages({
        ...errors,
        [id]: '',
      });
      inputElement.style.border = '1px solid #DAE2E7';
    }
  };

  const onCustomerContactSelect = (e: React.ChangeEvent<HTMLSelectElement>) => {
    const selectedContactId = e.currentTarget.value;
    const selectedContact = customerContacts.filter((obj: models.CustomerDetailsContactsGroupItem) => {
      return obj.contactDC === selectedContactId;
    });
    dispatch(setOpportunityContacts(selectedContact));
  };

  React.useEffect(() => {
    if (opportunity?.desc && opportunity?.oppRecordType && opportunity?.customer && opportunity?.stage) {
      setNextButtonEnabled(false);
    } else {
      setNextButtonEnabled(true);
    }
  }, [opportunity]);

  React.useEffect(() => {
    if (!state.enviornmentConfigs.crmOpportunityTypes.length) {
      dispatch(getOpportunityTypes());
    }
    const tempOpptyObject = { ...state.addOpportunity.opportunityDefaultParams };
    if (tempOpptyObject?.oppRecordType) {
      selectOpportunityType(tempOpptyObject.oppRecordType);
    }
    if (!tempOpptyObject.stage) {
      tempOpptyObject.stage = state.enviornmentConfigs.defaultOpprtunityInfo.stageCreated;
    }
    if (!state.enviornmentConfigs.crmOpportunityStage.length) {
      dispatch(saveOpportunityStages());
    }

    if (tempOpptyObject.customer) {
      loadCustomerContacts(tempOpptyObject.customer);
    } else if (contextValue?.customerId) {
      tempOpptyObject.customer = contextValue.customerId;
      tempOpptyObject.customerName = contextValue.customerName;
      loadCustomerContacts(contextValue.customerId);
    }
    setOpportunityField(tempOpptyObject);
    dispatch(saveOpportunityParams(tempOpptyObject));
  }, []);

  return (
    <>
      <div className="opportunity-step-circles">
        <ul className="list-inline step-circles">
          <li className="list-inline-item circle-stepone steps active">
            <span className="num">1</span>
          </li>
          <li className="list-inline-item circle-steptwo steps" role="presentation" onClick={() => onClickStep2()}>
            <span className="num">2</span>
          </li>
          <li className="list-inline-item circle-stepthree steps" role="presentation" onClick={() => onClickStep3()}>
            <span className="num">3</span>
          </li>
        </ul>
      </div>
      <div className="opportunity-forms form-top">
        <p className="stepone-title">{i18n.t('addOpportunityTitle')}</p>
        {state.enviornmentConfigs.crmOpportunityTypes.length && state.enviornmentConfigs.crmOpportunityStage.length ? (
          <div className="">
            <div className="steps-one-forms">
              <form>
                <div className="form-group oppty-form-elements">
                  <label htmlFor="desc" className="opp-label">
                    {i18n.t('addOpportunityNameLabel')}
                  </label>
                  <input
                    type="text"
                    className="form-control"
                    value={opportunity?.desc}
                    placeholder="Give opportunity a name"
                    id="desc"
                    onBlur={onBlur}
                    onChange={onInputValueChange}
                  />
                  <span className="form-hints">{errors?.desc}</span>
                </div>
                <div className="form-group oppty-form-elements">
                  <label htmlFor="customer" className="opp-label">
                    {i18n.t('addOpportunitySelectCustomerLabel')}
                  </label>
                  {contextValue?.customerId && contextValue?.customerName ? (
                    <input
                      type="text"
                      id="customer"
                      className="form-control"
                      placeholder=""
                      contentEditable={false}
                      onBlur={onBlur}
                      disabled={!!(contextValue && contextValue.customerId)}
                      value={opportunity?.customerName}
                    />
                  ) : (
                    <AsynSearchInput id="customer" onSearch={searchCustomers} onSearchItemSelect={onSearchItemSelect} />
                  )}
                  <span className="form-hints">{errors?.customer}</span>
                </div>

                <div className="form-group oppty-form-elements">
                  <label htmlFor="customer-contact" className="opp-label">
                    {i18n.t('addOpportunityAddCustomerContact')} <span className="opt-field">({i18n.t('addOpportunityOptionalField')})</span>
                  </label>
                  <select className="form-control iptor-dd" id="customer-contact" onChange={onCustomerContactSelect}>
                    <option selected>{i18n.t('addOpportunitySelectCustomerContact')}</option>
                    {customerContacts.map((obj: models.CustomerDetailsContactsGroupItem) => {
                      if (selectedContacts.length && selectedContacts[0].contactDC === obj.contactDC) {
                        return (
                          <option selected value={obj.contactDC}>
                            {obj.contactPerson}
                          </option>
                        );
                      }
                      return <option value={obj.contactDC}>{obj.contactPerson}</option>;
                    })}
                  </select>
                </div>

                <div className="form-group oppty-form-elements">
                  <label htmlFor="stage" className="opp-label">
                    {i18n.t('addOpportunitySelectStage')}
                  </label>
                  <select className="form-control iptor-dd" id="stage" onChange={onInputValueChange} value={opportunity?.stage}>
                    {state.enviornmentConfigs.crmOpportunityStage.map((obj: models.StageInfo) => {
                      return (
                        <option value={obj.salesStage} selected={state.enviornmentConfigs.defaultOpprtunityInfo.stageCreated === obj.salesStage}>
                          {obj.salesStage} - {obj.description}
                        </option>
                      );
                    })}
                  </select>
                  <span className="form-hints">{errors?.stage}</span>
                </div>

                <div className="form-group oppty-form-elements">
                  <label htmlFor="oppty-type" className="opp-label">
                    {i18n.t('addOpportunitySelectOpportunityType')}
                  </label>
                  <br />
                  <span className="form-hints">{errors?.stage}</span>
                  <div className="opportunity-type-container">
                    <OpportunityTypeList
                      opptyTypes={state.enviornmentConfigs.crmOpportunityTypes}
                      doClick={onOpportunityTypeSelect}
                      selected={selectedOpportunityType}
                    />
                  </div>
                </div>
              </form>
            </div>
          </div>
        ) : (
          <p>{i18n.t('formNotLoading')}</p>
        )}
      </div>
      <div className="step-nextbtn-with-arrow stepsone-nxtbtn">
        <button
          type="button"
          disabled={nextButtonEnabled}
          className={nextButtonEnabled ? 'stepone-next-btn inactive' : 'stepone-next-btn '}
          onClick={onNextButtonClick}>
          {i18n.t('nextButton')}
          <span className="right-whit-arrow">
            <img src={ImageConfig.CHEVRON_RIGHT_WHITE} alt="Next Arrow" />
          </span>
        </button>
      </div>
    </>
  );
};

export default AddOpportunityDefaultFields;
