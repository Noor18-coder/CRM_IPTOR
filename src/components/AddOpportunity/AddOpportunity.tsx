import React from 'react';
import { Dispatch } from 'redux';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router';
import { isArray } from 'lodash';
import Alert from 'react-bootstrap/Alert';
import AddOpportunityUserDefinedFields from './AddOpportunityUserDefinedFields';
import AddOpportunityDefaultFields from './AddOpportunityDefaultFields';
import AddOpportunitySelectItems from './AddOpportunitySelectItems';
import ImageConfig from '../../config/ImageConfig';
import { Constants } from '../../config/Constants';

import { AppState } from '../../store/store';
import {
  setOpportunityWindowActive,
  setOpportunityLoader,
  resetOpportunityData,
  setAddOpportunityError,
  setAddOpportunitySuccess,
} from '../../store/AddOpportunity/Actions';

import AddOpportunityApi from '../../helpers/Api/AddOpportunityApi';
import { Attributes } from '../../helpers/Api/Attributes';
import * as models from '../../helpers/Api/models';
import i18n from '../../i18n';

const AddOpportunity: React.FC = () => {
  const [step, setStep] = React.useState<number>(1);
  const state: AppState = useSelector((appState: AppState) => appState);
  const history = useHistory();
  const dispatch: Dispatch<any> = useDispatch();

  const createOpportunity = async () => {
    const { attributes, contacts, opportunityDefaultParams, products } = state.addOpportunity;
    opportunityDefaultParams.handler = state.auth.user.user;
    dispatch(setOpportunityLoader(true));
    const data = await AddOpportunityApi.add(opportunityDefaultParams);

    if (data && data.data && data.data.opportunityId) {
      const opptyId = data.data.opportunityId;
      Promise.all(
        attributes.map((obj: models.SaveAttributeFieldParam) => {
          obj.parentFile = Constants.OPPORTUNITY_ATTRIBUTES_PARENT_FILE;
          obj.parentId = opptyId;

          const params: models.SaveAttributeFieldParam = {
            parentFile: Constants.OPPORTUNITY_ATTRIBUTES_PARENT_FILE,
            parentId: opptyId,
            attributeType: obj.attributeType,
          };

          if (obj.attributeValueD) {
            params.attributeValueD = obj.attributeValueD;
          } else {
            params.attributeValue = obj.attributeValue;
          }
          return Attributes.addAttribute(params);
        })
      ).then((response) => {
        return response;
      });

      Promise.all(
        products.map((item: models.Item) => {
          return AddOpportunityApi.addItem(opptyId, item.item, 1, item.stockingUnit);
        })
      ).then((response) => {
        return response;
      });

      Promise.all(
        contacts.map((contact: models.CustomerDetailsContactsGroupItem) => {
          const params: models.AddCustomerContactParams = {
            contactParentId: opptyId,
            contactPerson: contact.contactPerson,
            phone: contact.phone,
            mobile: '',
            linkedin: '',
            fax: contact.fax,
            email: contact.email,
            contactDC: contact.contactDC,
          };
          return AddOpportunityApi.addContact(params);
        })
      ).then((response) => {
        return response;
      });

      dispatch(setOpportunityWindowActive(false));
      dispatch(setOpportunityLoader(false));

      dispatch(resetOpportunityData());
      if (opptyId) {
        history.push({ pathname: '/opp-details', state: { oppid: opptyId } });
      }
      return dispatch(setAddOpportunitySuccess(true));
    } else {
      dispatch(setOpportunityLoader(false));
      if (data.messages && isArray(data.messages) && data.messages[0] && data.messages[0].text) {
        return dispatch(setAddOpportunityError(data.messages[0].text));
      } else {
        return dispatch(setAddOpportunityError('Error occured while creating new opportunity.'));
      }
    }
  };

  const closeAction = () => {
    dispatch(resetOpportunityData());
    dispatch(setOpportunityWindowActive(false));
    dispatch(setOpportunityLoader(false));
  };

  const changeStep = (num: number) => {
    setStep(num);
  };

  const hideErrorMessage = () => {
    dispatch(setAddOpportunityError(''));
  };

  return (
    <>
      <div className="sliding-panel-container fixed-header-footer">
        <div className="sliding-panel">
          <div className="title-row opp-header-text">
            <button type="button" className="mob-steps-back" onClick={closeAction}>
              <img src={ImageConfig.CHEVRON_LEFT} alt="Back" />
            </button>
            {i18n.t('addNewOppt')}
            <button type="button" className="panel-close-icon link-anchor-button" onClick={closeAction}>
              <img src={ImageConfig.CLOSE_BTN} alt="Close" />
            </button>
          </div>
          {state.addOpportunity.error ? (
            <Alert variant="danger" onClose={hideErrorMessage} dismissible>
              <Alert.Heading>Error</Alert.Heading>
              <p>{state.addOpportunity.error}</p>
            </Alert>
          ) : null}
          <div className="all-opportunity-steps-container">
            {step === 1 ? <AddOpportunityDefaultFields changeStep={changeStep} /> : null}
            {step === 2 ? <AddOpportunityUserDefinedFields changeStep={changeStep} /> : null}
            {step === 3 ? <AddOpportunitySelectItems changeStep={changeStep} createOpportunity={createOpportunity} /> : null}
          </div>
        </div>
      </div>
    </>
  );
};

export default AddOpportunity;
