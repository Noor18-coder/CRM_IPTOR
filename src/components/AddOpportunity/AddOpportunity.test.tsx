import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import AddOpportunity from './AddOpportunity';
import StoreMock from '../../mocks/Store.mock';

const store = StoreMock.createInitialStore();

describe('[AddOpportunity] AddOpportunity', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <AddOpportunity />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
