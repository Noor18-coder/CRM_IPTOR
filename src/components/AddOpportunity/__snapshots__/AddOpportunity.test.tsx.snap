// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[AddOpportunity] AddOpportunity should renders correctly 1`] = `
<div
  className="sliding-panel-container fixed-header-footer"
>
  <div
    className="sliding-panel"
  >
    <div
      className="title-row opp-header-text"
    >
      <button
        className="mob-steps-back"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="Back"
          src="chevron-left-black.svg"
        />
      </button>
      Add New Opportunity
      <button
        className="panel-close-icon link-anchor-button"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="Close"
          src="close.svg"
        />
      </button>
    </div>
    <div
      className="all-opportunity-steps-container"
    >
      <div
        className="opportunity-step-circles"
      >
        <ul
          className="list-inline step-circles"
        >
          <li
            className="list-inline-item circle-stepone steps active"
          >
            <span
              className="num"
            >
              1
            </span>
          </li>
          <li
            className="list-inline-item circle-steptwo steps"
            onClick={[Function]}
            role="presentation"
          >
            <span
              className="num"
            >
              2
            </span>
          </li>
          <li
            className="list-inline-item circle-stepthree steps"
            onClick={[Function]}
            role="presentation"
          >
            <span
              className="num"
            >
              3
            </span>
          </li>
        </ul>
      </div>
      <div
        className="opportunity-forms form-top"
      >
        <p
          className="stepone-title"
        >
          Opportunity Name & Type
        </p>
        <p>
          Sorry, we are not able to load form at the moment. Please try later.
        </p>
      </div>
      <div
        className="step-nextbtn-with-arrow stepsone-nxtbtn"
      >
        <button
          className="stepone-next-btn inactive"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          NEXT
          <span
            className="right-whit-arrow"
          >
            <img
              alt="Next Arrow"
              src="chevron-right-white.svg"
            />
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
